src/sys/socket/mod.rs

```diff
@@ -1334,3 +1348,233 @@ pub fn shutdown(df: RawFd, how: Shutdown) -> Result<()> {
+    target_os = "freebsd",
+    target_os = "netbsd",
+    target_os = "android",
+))]
+impl<'a> RecvMMsgHdr<'a> {
+    pub fn new(iov: &'a mut[IoVec<&'a mut [u8]>],
+                  cmsg_buffer: Option<&'a mut CmsgBuffer>,
+                  flags: MsgFlags,
+                  addr: Option<&'a mut SockAddr>) -> RecvMMsgHdr<'a> {
+        let (name, namelen) = match addr {
+            Some(addr) => {
+                let (x, y) = unsafe { addr.as_ffi_pair_mut() };
+                (x as *mut libc::sockaddr, y)
+            },
+            None => (ptr::null_mut(), 0),
+        };
+        let (msg_control, msg_controllen) = match cmsg_buffer {
+            Some(cmsgspace) => {
+                let msg_buf = cmsgspace.as_bytes_mut();
+                (msg_buf.as_mut_ptr(), msg_buf.len())
+            },
+            None => (ptr::null_mut(), 0),
+        };
+        let mut hdr: libc::mmsghdr = unsafe { mem::uninitialized() };
+        hdr.msg_hdr.msg_control = msg_control as *mut _;
+        hdr.msg_hdr.msg_controllen = msg_controllen as _;
+        hdr.msg_hdr.msg_flags = flags.bits();
+        hdr.msg_hdr.msg_iov = iov.as_mut_ptr() as *mut iovec;
+        hdr.msg_hdr.msg_iovlen = iov.len() as _;
+        hdr.msg_hdr.msg_name = name as _;
+        hdr.msg_hdr.msg_namelen = namelen;
+        hdr.msg_len = 0;
+        RecvMMsgHdr(hdr, PhantomData)
+    }
+
+    /// The number of bytes actually transferred, after a call to `sendmmsg()` or `recvmmsg()`.
+    pub fn msg_len(&self) -> usize {
+        self.0.msg_len as usize
+    }
+
+    pub fn address(&self) -> Result<SockAddr> {
+        unsafe { sockaddr_storage_to_addr(&*(self.0.msg_hdr.msg_name as *const libc::sockaddr_storage),
+                                          self.0.msg_hdr.msg_namelen as usize) }
+    }
+
+    pub fn cmsgs(&self) -> CmsgIterator {
+        let cmsghdr = unsafe {
+            if self.0.msg_hdr.msg_controllen > 0 {
+                // got control message(s)
+                debug_assert!(!self.0.msg_hdr.msg_control.is_null());
+                CMSG_FIRSTHDR(&self.0.msg_hdr as *const msghdr)
+            } else {
+                ptr::null()
+            }.as_ref()
+        };
+        CmsgIterator {
+            cmsghdr,
+            mhdr: &self.0.msg_hdr,
+        }
+    }
+}
+
+/// Receive multiple messages from a socket using a single system call.
+#[cfg(any(
+    target_os = "linux",
+    target_os = "freebsd",
+    target_os = "netbsd",
+    target_os = "android",
+))]
+pub fn recvmmsg(fd: RawFd, msgvec: &mut[RecvMMsgHdr],
+                flags: MsgFlags,
+                mut timeout: Option<TimeSpec>) -> Result<usize> {
+    let tptr = match timeout {
+        Some(ref mut time) => time.as_ref() as *const libc::timespec,
+        None => ptr::null_mut(),
+    };
+    let ret = unsafe {
+        libc::recvmmsg(
+            fd,
+            msgvec.as_mut_ptr() as *mut libc::mmsghdr,
+            msgvec.len() as _,
+            flags.bits(),
```

Comment:
Note: on `musl` the `flag` argument is an `unsigned int`, while on `gnu`, it is a signed int.